deployer/aws-lambda/content-origin-request/index.js

```diff
@@ -50,43 +61,85 @@ exports.handler = async (event) => {
+  // Do we need to insert the locale? If we do, trim a trailing slash
+  // to avoid a double redirect, except when requesting the home page.
   if (
-    request.uri.startsWith("/docs/") ||
+    request.uri === "" ||
     request.uri === "/" ||
-    request.uri === ""
+    NEEDS_LOCALE.test(requestURILowerCase)
   ) {
     const path = request.uri.endsWith("/")
       ? request.uri.slice(0, -1)
       : request.uri;
     const locale = getLocale(request);
     // The only time we actually want a trailing slash is when the URL is just
     // the locale. E.g. `/en-US/` (not `/en-US`)
-    return redirect(`/${locale}${path || "/"}`);
+    return redirect(`/${locale}${path || "/"}` + qs);
   }
 
-  // A document URL with a trailing slash should redirect
-  // to the same URL without the trailing slash.
+  // At this point, the URI is guaranteed to start with a forward slash.
+  const uriParts = request.uri.split("/");
+  const uriFirstPart = uriParts[1];
+  const uriFirstPartLC = uriFirstPart.toLowerCase();
+
+  // Do we need to redirect to the properly-cased locale? We also ensure
+  // here that requests for the home page have a trailing slash, while
+  // all others do not.
   if (
-    request.uri.endsWith("/") &&
-    request.uri.toLowerCase().includes("/docs/")
+    VALID_LOCALES.has(uriFirstPartLC) &&
+    uriFirstPart !== VALID_LOCALES.get(uriFirstPartLC)
   ) {
-    return redirect(request.uri.slice(0, -1), {
+    // Assemble the rest of the path without a trailing slash.
+    const extra = uriParts.slice(2).filter(Boolean).join("/");
+    return redirect(`/${VALID_LOCALES.get(uriFirstPartLC)}/${extra}${qs}`);
+  }
+
+  // Handle cases related to the presence or absence of a trailing-slash.
+  if (LOCALE_URI_WITHOUT_TRAILING_SLASH.has(requestURILowerCase)) {
+    // Home page requests are the special case on MDN. They should
+    // always have a trailing slash. So a home page URL without a
+    // trailing slash should redirect to the same URL with a
+    // trailing slash. When the redirected home-page request is
+    // processed by this Lambda function, note that we'll remove
+    // the trailing slash before the request reaches S3 (see below).
+    return redirect(request.uri + "/" + qs, {
       status: 301,
-      cacheControlSeconds: 3600 * 24 * 30,
+      cacheControlSeconds: THIRTY_DAYS,
+    });
+  } else if (LOCALE_URI_WITH_TRAILING_SLASH.has(requestURILowerCase)) {
+    // We've received a proper request for a locale's home page (i.e.,
+    // it has a traling slash), but since that request will be served
+    // from S3, we need to strip the trailing slash before it reaches
+    // S3. This is required because we store the home pages in S3 as
+    // their path name itself, for example "en-us" for the English home
+    // page, not "en-us/index.html", which is what S3 would look for if
+    // we left the trailing slash.
+    request.uri = request.uri.slice(0, -1);
+  } else if (request.uri.endsWith("/")) {
```

Comment:
We can't do this to requests for `/{locale}/account/` since that endpoint (in Kuma) **requires** a trailing slash. If we do, we get a redirect loop, since Kuma redirects `/{locale}/account` to `/{locale}/account/`. 🤦

Reply:
I'm out of time to solve this today. There are several options available, including routing account requests directly to Kuma, bypassing lambda@edge (but that requires at least 3 new CDN behaviors to handle all of the cases).